### ISSUE TITLE
Developer Experience: There's ~280 __experimental APIs. Let's stabilize them!

### ISSUE THREAD

**Issue by adamziel** [WordPress/gutenberg#40316]
## The problem

[Rob's script](https://github.com/WordPress/gutenberg/issues/28157) detectes the following number of `__experimental` APIs:

- Gutenberg 5.0 has 81 `__experimental` APIs
- Gutenberg 6.0 has 113 `__experimental` APIs
- Gutenberg 7.0 has 173 `__experimental` APIs
- Gutenberg 8.0 has 224 `__experimental` APIs
- Gutenberg 9.0 has 300 `__experimental` APIs
- Gutenberg 10.0 has 337 `__experimental` APIs
- Gutenberg 11.0 has 442 `__experimental` APIs
- Gutenberg 12.0 has 562 `__experimental` APIs
- Gutenberg 13.0 has 614 `__experimental` APIs

It is a generous estimate and, as @talldan pointed, the actual number is 280 as of Gutenberg 13.0. Still, the progression above demonstrates the problem.

The [Audit Experimental APIs WP 6.0](https://github.com/WordPress/gutenberg/issues/39977) issue I've created for WordPress 6.0 initially exceeded GitHub's characters limit:

![](https://user-images.githubusercontent.com/205419/161272472-31bc094d-616f-4dd2-aadd-fc9cbc178c4f.png)

And that's just the `__experimental` ones! There are also `__unstable`, `__internal`, and probably other prefixes..

### So what?

Plugin and theme authors are forced to rely on the `__experimental` features that could get removed or changed in a backwards incompatible way at any time. It is a serious maintenance burden. Every new Gutenberg/WordPress release means potentially breaking changes. 

Also, shipping essential features as experimental sends the wrong message to the developers: _With all these `__experimental` APIs around, adding another one surely can't be a big deal._ What does "experimental" even mean at this point? What is a valid use case, and what isn't?

## The discussion

Let's find a way to harness this growth!

There are two parts of this:

- Stabilize the existing __experimental APIs
- Create less new __experimental APIs

### Stabilize the existing __experimental APIs

Many are de-facto stable, we just never marked them as such.

One idea to get there would be to:
* Automatically create one PR per __experimental API to stabilize it
* Ping the original author and the reviewers
* Perhaps even publish a call for stabilization on make.wordpress.org
* Let it simmer for a few weeks

Some PRs would get closed, some ignored, and others merged. Once that happens, we regroup and figure it out from there.

Upsides:
- Each __experimental API would get feedback (even silence provides information)
- Many __experimental APIs would likely get promoted

Downsides:
- A lot of noise
- Some __experimental APIs might be intertwined and would have to be promoted as a bundle

To me, that's a worthy trade-off. Is it to you? What other ways do you see? 

Side note: Plugins relying on these "permanently __experimental" names would break – these names would have to keep working with a deprecation notice.

### Create less new __experimental APIs

The number will continue to grow until something slows it down.

It could be a project policy, but policies need enforcement. It is hard, creates extra work, and may feel discouraging on the receiving end. 

How about a soft nudge? Perhaps a CI check or a bot that detects new __experimental APIs and tries to educate the PR author? Ideally, it would also propose an alternative. 

cc @gziolo @mamaduka @noisysocks @mtias @youknowriad @andronocean @ciampo @aaronrobertshaw @c4rl0sbr4v0 @jorgefilipecosta @mcfs @paaljoachim @scruffian @draganescu @getdave @talldan @oandregal @ntsekouras @ellatrix @ndiego @priethor @annezazu @dmsnell

**Comment by gziolo** [comment]
> Automatically create one PR per __experimental API to stabilize it

Interesting idea. What possibly could go wrong? 😃

I think the biggest concern here would be the documentation which might be missing in many cases. The automated PR could serve as s good way to find out how many undocumented APIs we have. I bet that the lack of good explanation of how to use those APIs is the biggest issue because, at this time, extenders probably learned how to work with experimental prefix 😅

**Comment by talldan** [comment]
> Gutenberg 13.0 has 614 __experimental APIs

The numbers aren't quite right as that script lists many duplicates. It shows any existence of `__experimental`. It doesn't only list where those experimental APIs are defined, also where they're used.

Doing some quick sorting and removing duplicates gives a number around 280. It might be a little higher because there are probably some APIs that are actually defined in multiple places with the same name.

Still quite high. I think the tricky thing is prioritizing the ones that are actually useful to developers.

The age of an experimental API is also an interesting measurement. If an API is going to sit around for years as experimental, perhaps it'd be better being switched to `__internal`.

edit: sorry for closing and reopening accidentally.

**Comment by paaljoachim** [comment]
Some brain storming...
It seems this topic needs a wider discussion. If it has not been brought up earlier in the core editor meeting, it could be brought up there. Posting a comment about this to the agenda so people will notice that it will be discussed during the meeting and then spending part of the Open Floor having a wider discussion.

Another thought is to focus on stabilizing experimental APIs for a 6.x version where main focus is stabilization. It would then need to be posted on make core well ahead of time to get developers ready for the shift from experimentation to stabilization.

**Comment by dmsnell** [comment]
when I recently added `__unstableBlockSource` I did so because there wasn't clear guidance on intentionally-internal APIs. had there been, and had there been documentation, I would have used `__internal`.

notable [we don't list `__internal`](https://github.com/WordPress/gutenberg/blob/f4a031d1af9fd196a6cb1e0c5172da658431221d/docs/contributors/code/coding-guidelines.md#experimental-and-unstable-apis) as an official prefix; if we want to draw a distinction there between `__unstable` and `__internal` we should consider making that decision and documenting it first.

**Comment by mtias** [comment]
Can we approach this in a more gradual manner from oldest to newest? Doesn't make sense to treat APIs introduced in the last couple releases the same as those introduced a couple years ago. APIs that are already part of major releases should be prioritized.

> Many are de-facto stable, we just never marked them as such.

Which ones? Identifying those seems like a good start to me.

**Comment by noisysocks** [comment]
Getting the age of these tokens would be useful. Could probably hack my script to do this using `git log --oneline -S $TOKEN_NAME`. I'd say anything that's been in the codebase for more than 12 months should be marked stable or `__internal`.

**Comment by peterwilsoncc** [comment]
To share my thoughts from Slack in a less ephemeral form.

WordPress Core is considered the stable version of the block editor. It has a long standing commitment.  to maintaining backward compatibility of public APIs. Any of the `__` prefixed items that have been exported from packages in WP Core will need to be maintained.

Discussing how the block editor is used by developers, in order to provide a native feeling experience plugins are required to use the private functions. This is reflected in a rudimentary [search of the plugins directory](https://wpdirectory.net/search/01G1PNVVD3Q9CJ5K41KM33F8C2) and [themes directory](https://wpdirectory.net/search/01G23SVVJBQQ7D0X5VZRYCZRQW).

Maintaining deprecated private APIs has been a long standing practice of WordPress, as reflected in [deprecated.php](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/deprecated.php).

Basically, hard deprecating in WordPress Core is a no-go. Instability between versions is already beginning to alienate some of the block-editors biggest external advocates so a concerted effort of deprecation would simply expedite the alienation. 

As Gutenberg is considered a bleeding edge product, hard deprecation is acceptable. I'd suggest a soft deprecation of a short period of time (three to six months) before hard deprecation. If it doesn't already include it, adding the hard deprecation version number to soft deprecation console messages would be ideal.

**Comment by adamziel** [comment]
All great notes @peterwilsoncc! It's an interesting situation:

* We have APIs are called `__experimental` so they can be hard-deprecated
* Some of them made it to WordPress Core as public APIs so they cannot be hard-deprecated anymore

In other words, releasing an `__experimental` API as a public API in WordPress Core makes it de-facto not experimental. It's just the codebase doesn't reflect that reality. In this case, I'd like to make sure we either remove __experimental prefixes from new public APIs or turn these public and __experimental APIs into private ones before each WordPress core release.

**Comment by adamziel** [comment]
Actually, that doesn't seem to be possible. At least these two kinds of APIs cannot be made private before a major release:

* React properties
* experimental APIs exposed by package A to be used in package B

There must be a way to ship them and retain the ability to remove them. If anything that's made public cannot be easily removed, there's no point in maintaining Gutenberg as a separate repository. I now err on the side of keeping the ability to hard-deprecate experimental APIs even if they made it as public to WordPress core.

This is consistent with the following handbook page:

https://developer.wordpress.org/block-editor/contributors/code/coding-guidelines/#experimental-and-unstable-apis

We could, though, make sure that the __experimental prefix keeps working for some time after stabilizing the underlying API (cc @gziolo @dmsnell).

**Comment by Clorith** [comment]
I feel like @peterwilsoncc is much better at I than articulating here, but it should be noted that the handbook page you are referencing is the contributor guidelines for the Gutenberg plugin, and not the implementor guidelines for developers using what actually lands in core.

If the worry is that experimental features are needed for a feature to be put into core, would it then not stand to reason that that feature is not actually ready for core yet, because it is in part experimental? (I'm being intentionally direct here, I know there are nuances, but would this not improve developer experiences as well if not even core relied on experimental or unstable components?)

**Comment by noisysocks** [comment]
I believe that the intention of `__experimental` and `__unstable` was to mark an API as subject to breaking without warning _regardless_ of whether that API appears in Core or Gutenberg.

I know Core has historically maintained backwards compatibility for APIs marked private but I'd push back on that precedent a little bit as the majority of this has been for APIs written in PHP. Keeping PHP around for backwards compatibility carries a maintenance burden but not much else. Keeping JavaScript around for backwards compatibility, however, carries a maintenance burden **and a user experience burden** as it increases the bundle size.

**Comment by talldan** [comment]
> If the worry is that experimental features are needed for a feature to be put into core, would it then not stand to reason that that feature is not actually ready for core yet, because it is in part experimental?

I think it's worth dividing it up into two things—the user facing feature and the developer API. The user facing feature is usually ready for release, but developer APIs still somewhat open to change. It's often possible to change developer APIs with no effect on the user experience.

On the other hand 'ready' is a relative term. User interfaces do naturally change over time. It can be difficult to do that if the underlying APIs all require an endless commitment to backwards compatibility, and I think that's the main reason to consider something experimental.

**Comment by peterwilsoncc** [comment]
> I believe that the intention of `__experimental` and `__unstable` was to mark an API as subject to breaking without warning _regardless_ of whether that API appears in Core or Gutenberg.

I accept that was the intention but once these APIs started shipping in the stable product the promise to maintain backward compatibility was made. These APIs have only ever been documented as removed without notice in the plugin contributor guide.

* `__experimentalGetSettings` is used on 11 million plugin installs, the only way to get the sites date format
* `__experimentalLinkControl` is used on 11 million plugin installs, as Dave has covered elsewhere. It's the only way to introduce native equivalent UIs.
* `__experimentalNumberControl` Six million odd plugin installs,
* `__experimentalRadio` 500K plugin installs (excluding gutenberg)
* `__experimentalRadioGroup` 500K plugin installs (excluding Gutenberg)

This is before considering themes and custom builds in which use these items too. I am getting DMs since joining this ticket.

Yup, there's a maintenance burden to maintain backward compatibility. That's the cost of working on a project that strives to maintain it.

Yup, shipping these APIs in the stable product has introduced a user experience burden too. Realistically, the APIs should not have been shipped in an unstable/experimental form.

**Comment by noisysocks** [comment]
The current policy is that `__experimental` and `__unstable` APIs have no backwards compatibility guarantee.

https://github.com/WordPress/gutenberg/blob/336748b39cc1e079806c87c2b8f30ed32d0ebb26/docs/contributors/code/coding-guidelines.md#experimental-and-unstable-apis

We don't really have a way of picking and choosing which code lands from Gutenberg into Core which is a large part of why `__experimental` and `__unstable` were introduced. 

I would prefer that we improve how we communicate what our policy is for `__experimental` and `__unstable` APIs rather than change what that policy is. For example, maybe we can automatically output a `console.warn()` message if any client code calls an experimental API.

For experimental APIs that are very frequently used (e.g. the ones you list) I think it makes sense to be pragmatic and mark those APIs as deprecated for a few releases. Not _guaranteeing_ backwards compatibility doesn't mean that we can't do it anyway.

But I disagree completely that any `__experimental` or `__unstable` API becomes de facto stable once in Core. The team has held the exact opposite assumption for quite some time.

**Comment by mtias** [comment]
I think there can be a stronger commitment to not include `__experimental` and `__unstable` APIs in major releases, though as mentioned by @noisysocks that won't always be realistic. We should take those instances as practical occasions to force stabilization as much as possible. In any case, these APIs don't have a commitment to remain as is because they were designed to be changed (at the very least, to drop the prefix at some point) and there's a shared responsibility between the core project and extenders in how they interact with them.

`deprecated.php` is not really relevant as a comparison since server-side PHP deprecations don't burden users and visitors with degraded performance. It's not realistic for the project to bundle JS functions forever without some form of "present only if needed" mechanism, which is not entirely realistic as of this time. There's probably a need for a wider policy discussion there that balances expectations and pragmatism.

I also see two things at play here: the use and abuse of experimental APIs during the API design (generally to be used and tested in the Gutenberg plugin) and the lack of a diligent process for stabilizing them when they satisfy design criteria. The ones that are to be considered _de facto_ public are those that have existed for many releases in a stable form despite their nomenclature.

**Comment by peterwilsoncc** [comment]
@m @josephahaden  Has WordPress dropped the core philosphy to maintain backward compatibility? Is the project now willing to break millions of plugin installs per my [comment above](https://github.com/WordPress/gutenberg/issues/40316#issuecomment-1120635542)?

**Comment by mtias** [comment]
@peterwilsoncc pardon me, but that seems quite the leap. Where do you see the argument to break millions of plugin installs? There's a nuanced discussion to be had when it comes to how much code we are willing to ship to users on the client in the name of keeping APIs working that's collateral to the discussion of how experimental APIs are introduced, how they are refined, how they are stabilized, and what it means when these end up in major releases. Among those, there are some very concrete things to pursue that's what we should focus effort on.

**Comment by mtias** [comment]
I'll try to summarize what I see as more concrete items.

- Indefinitely increasing client-side code is at odds with performance budgets, we need to be cognizant of this reality. We might have more room when async loading and native imports can be leveraged.
- When designing APIs, feedback from real usage is crucial. WordPress beta periods are not long enough or timely enough to gather wide feedback. The bleeding edge Gutenberg plugin is more appropriate for it.
- Unstable APIs that leak need to be looked at explicitly at the time of major releases for stability as the commitment is orders of magnitude higher.
- Current experimental APIs bundled in core need to be addressed on a case by case basis with an aim to stabilize, probably from oldest to newest.
- Mistakes will be made and we need to design for iteration and change while not breaking things. This will always be challenging at heart and need ongoing ingenuity.

**Comment by ndiego** [comment]
As a plugin author that is currently using _many_ `__experimental` APIs, I would love to see these stabilized. Most provide crucial functionality but building a product that relies on an `__experimental` API is always a bit disconcerting. So long as the process is exceedingly transparent, is well publicized, and we provide plugin/theme authors with a guide on how to migrate to stable versions, then I like this initiative.

**Comment by dmsnell** [comment]
> and a user experience burden

in my opinion crashing the editor or a block is a much bigger user experience burden than adding _a few extra tens of gzip'd bytes_ which is what the cost is going to amount to by leaving in deprecation notices on the old function names.

> indefinitely increasing client-side code

this also feels like quite the leap @mtias. we're discussing ~280 functions and a rename of those functions.

when we mark code as deprecated we don't really leave clues in the source itself on when to pull them out, but I wonder about this, if we decided a deprecation policy of six months or so and leave some kind of `removeAfter` property to make it more obvious to someone passing by

```ts
export const ButtonBlockerAppender = forwardRef( ( props, ref ) => {
	deprecated( `wp.blockEditor.ButtonBlockerAppender`, {
		alternative: 'wp.blockEditor.ButtonBlockAppender',
		since: '5.9',
		removeAfter: '2022/01/01'
	} );

	return ButtonBlockAppender( props, ref );
} );
```

it just seems that overall a few months transition window (or one in which we can monitor usage in real plugins) that adds almost no cost to the generated build but prevents the editor from crashing (or a block from crashing) is an obvious choice to make.

> maybe we can automatically output a console.warn() message

Just a reminder @noisysocks that `console` usage in Gutenberg is really frustrating to deal with outside of WordPress. I know we don't have a better solution in the project right now, but I'll at least encourage everyone to avoid using `console` as much as possible since it doesn't play well with others.

**Comment by dmsnell** [comment]
should we start this work by purging old `deprecated()` functions?

**Comment by peterwilsoncc** [comment]
@mtias 

I don't think it's a leap because it's already happening and the tone of this discussion is that it's about to happen a lot more.

In my personal experience:

* unintentional backward compatibility breaks of stable, public APIs are not fixed, see https://github.com/WordPress/gutenberg/issues/38361 and the [backward compatibility label](https://github.com/WordPress/gutenberg/issues?q=is%3Aopen+is%3Aissue+label%3A%22Backwards+Compatibility%22)
* [snowflake code needs to be added](https://github.com/WordPress/wordpress-develop/blob/e592e3303a75a313b2938c27cf5b4fa1ad321bfa/src/wp-admin/includes/update-core.php#L1709-L1727) to core to deal with compatibility breaks with the WP itself. Not insignificantly, this doesn't work on sites using the plugin as a mu-plugin which is typical of enterprise sites
* the [intention to break backward compatibility](https://make.wordpress.org/core/2022/05/03/block-editor-miscellaneous-dev-notes-for-wordpress-6-0/) of stable CSS variables posted in a dev note without consultation of WP Core committers
* having to put conditional code in mini-plugins to account for backward compatibility breaks between versions 
* having to argue against backward compatibility breaks that come from plugins copying code from Gutenberg/WordPress core.

Having sort feedback from others

* core committers unwilling to use certain library features to make core tasks easier because they don't trust the reliability
* developers no longer upgrading WP client sites. As a core committer who has strived to maintain backward compatibility for years this disappoints me. As a security team member it's greatly concerning
* developers deciding to include copies of packages in themes and plugins rather than rely on the `wp.*` globals. Again this concerns me from a security perspective but it also increases the JavaScript payload significantly more than maintaining backward compatibility
* reports of backward compatibility breaks in minor versions: "you don't expect a 5.9.1 release to break the responsiveness of a bunch of images in our sites outside the block editor"
* developers considering never using core blocks because they're too unstable: "I stopped using/extending core blocks because they were changing too much and I’ve been using ACF Blocks so that I at least I know I can make blocks that won’t break. Granted the UI isn’t as good as core blocks but I’ll take stability over blocks breaking any time."

This is just from limited sources of feedback where I know people will treat the request for experiences with respect. I've not asked on places like twitter as it would quickly become a bin fire.

The documentation about the experimental APIs is hidden away in the contributor handbook. Even if it weren't, when an experimental API remains largely unchanged since it's introduction in WordPress 5.0 then it's reasonable developers consider it stable.

Trust me, I'd really rather not be burning political capital on this discussion. For the 21 releases of WordPress I've worked on, I have worked really hard to maintain backward compatibility and I feel like this work is been undermined on a consistent basis. I know I am not the only person that feels like this.

Furthermore, a lot of the discussion around these decisions happens in spaces I don't have access to as a Core committer.

**Solving the problem**

No further experimental APIs introduced to core. A rethink of the approach is needed because it's been used as a get out jail free card.

When building the `@wordpress/*` packages two modes need to be introduced: `stable` and `bleeding-edge`. The latter can be used in the plugin to expose new experimental APIs.

**Comment by noisysocks** [comment]
> When building the `@wordpress/*` packages two modes need to be introduced: `stable` and `bleeding-edge`. The latter can be used in the plugin to expose new experimental APIs.

We can't put experimental functions and components behind a flag (e.g. `bleeding-edge` like you suggest or the existing `process.env.IS_GUTENBERG_PLUGIN` flag) because it's often the case that an experimental component or function is used by a new feature that we want to release in Core. We need the code to be there for use by the Core feature but not used by third parties.

This is what the `__experimental` and `__unstable` prefix aims to address. It's supposed to scare the third parties away.

One idea maybe worth investigating: Could we make it so that experimental functions and components can only be accessed via an `import ... from '@wordpress/...';` statement? That is, they won't appear in the `wp` global. This is already how WordPress code is written so it shouldn't affect Core features. Adventurous third parties can use the experimental APIs but only by adding the `@wordpress` package to their `package.json` and therefore vendoring the code. When the experimental code is eventually deleted from Core it will continue to exist in the plugin's `node_modules` until the plugin is updated.

**Comment by luisherranz** [comment]
> Solving the problem
>
> No further experimental APIs introduced to core. A rethink of the approach is needed because it's been used as a get out jail free card.

Unfortunately, it's not that simple.

Typically, teams ship either:

- A product that requires a fast iteration of the UX using a stable DX tool.
  _They improve the UX and recover from previous mistakes by shipping new product versions._
- A DX tool that remains unstable until a major version is released.
  _They improve the DX and recover from previous mistakes by starting a new major version where they can introduce breaking changes again._

But Gutenberg contributors have been shipping at the same time:

- A product that requires fast iteration of the UX and backward compatibility with all existing WordPress sites.
- A DX tool that is required to have infinite backward compatibility, support the highest degree of extensibility of the market, needs to support the fast pace of the UX iteration, has backward compatibility with existing WordPress themes and doesn't have any system to recover from previous mistakes.

Please, if something, at least **let's acknowledge that what Gutenberg contributors have been doing is extremely complex**. Probably one of the most complex open-source projects in history in this regard.

Good design (UX or DX) needs iteration and consolidation. Gutenberg is both a UX product and a DX tool at the same time. **And DX tools also need iteration and consolidation**.

Experimental APIs are their (our) only chance of shipping new UX features before completely figuring out how the DX tool needs to be. Imagine where Gutenberg would be today if we had to get stuck with the very first UX that Gutenberg ever shipped.

> when an experimental API remains largely unchanged since it's introduction in WordPress 5.0 then it's reasonable developers consider it stable.

Unfortunately, that is not that simple either. React is an excellent example of this:

- They exposed experimental APIs for the concurrent mode three (maybe four?) years ago but kept iterating them until they stabilized them in React 18. _They kept testing these APIs on Facebook.com during all those years._
- They exposed experimental APIs (beta?) for hooks and stabilized them within a few months.
- They exposed experimental APIs for Server Components a year and a half ago. _They are still not stabilized._

I haven't seen any single React developer demanding the stabilization of the Concurrent or Server Components APIs because "they have been exposed for X years already".

We all agree that experimental APIs can't stay like that forever, **but experimental APIs need to be treated case by case**. I'm sure there will be cases where Gutenberg contributors have usage of experimental APIs has been unnecessary or incorrect, and other times where someone has regretted not using an experimental API. But we cannot put them all in the same bucket.

---

The only solution is to find a good compromise and improve our processes, including:

- Improve our documentation and communication to ensure extenders understand the possible consequences of using experimental APIs and align their expectations with those of Core contributors.
- Work with plugins that currently rely on the existing experimental APIs to figure out the solution for the current situation.
- Improve our processes. @mtias [summary is a great starting point](https://github.com/WordPress/gutenberg/issues/40316#issuecomment-1122177204) that contains actionable items.

Anything else is unrealistic to me.

**Comment by peterwilsoncc** [comment]
> One idea maybe worth investigating: Could we make it so that experimental functions and components can only be accessed via an `import ... from '@wordpress/...';` statement? That is, they won't appear in the `wp` global. This is already how WordPress code is written so it shouldn't affect Core features.

Thanks Rob, this is what I was trying (but failing) to suggest. The word I was looking for is _exposed_ rather than exported but I confused up my JS terminology.

A similar approach of locking away any APIs was taken for supporting [webfonts in theme.json](https://github.com/WordPress/wordpress-develop/blob/0907c144a86140197e242ea5bba2443f035a52c9/src/wp-includes/script-loader.php#L3058-L3529). A few of us worked on the ticket and didn't want to provide a developer API.

> Please, if something, at least **let's acknowledge that what Gutenberg contributors have been doing is extremely complex**. Probably one of the most complex open-source projects in history in this regard.

I do understand that, I really do. 

For an in-house project these conversations would happen in the room, possibly over a nice lunch. The nature of open-source is that even the room is public.

> Experimental APIs are their (our) only chance of shipping new UX features before completely figuring out how the DX tool needs to be. Imagine where Gutenberg would be today if we had to get stuck with the very first UX that Gutenberg ever shipped.

I'm not trying to block experimental APIs entirely, I'm just trying to limit them to the bleeding edge product. The best analogy I can think of is [CSS vendor prefixes](https://developer.mozilla.org/en-US/docs/Glossary/Vendor_Prefix): browsers intended to remove them but the end result was they had to [support each others](https://jsbin.com/viqocup/edit?html,css,output).

> * Current experimental APIs bundled in core need to be addressed on a case by case basis with an aim to stabilize, probably from oldest to newest.

This makes sense as a starting point. 

While searching the plugin repo is a good guide to how frequently something is used, I am also reaching out to people at various agencies to try and find out general guides about in house use. 

The types of sites measured in the "40%+ of the internet" number that's often quoted are usually custom builds so reaching out to plugin devs alone isn't enough. If Jetpack or Yoast is using something, there's a high chance the custom builld behind big-corporate-site-dot-com is using it too.

**Comment by peterwilsoncc** [comment]
To continue this discussion, I have to say I am a little annoyed to be told I am making a leap to presume there will be backward compatibility breaks only to discover that #40052 was intended to be merged last week unless someone objected. https://github.com/WordPress/gutenberg/pull/40052#issuecomment-1130103017

@Clorith had already objected on the pull request and I was pretty clear in this discussion that I objected to such breaking changed too. It's provocative and very disappointing. 

When experimental APIs were introduced in the Gutenberg plugin, the make/core post assured developers it was only for the bleeding edge product

> This strategy won’t be applied in Core because it’s very important that we continue to maintain the user trust with auto-updates and **there’s no intention in changing that in the future**.

-- _Original emphasis_, [source](https://make.wordpress.org/core/2018/11/14/technical-organisation-post-5-0/)

I created #41183 with a soft deprecation: the total package size increase was 45B. Napkin maths for maintaining compatibility is 10-15KB on a current bundle size of 4.79 MB for the post editor.

To repeat myself from an earlier comment, the loss of faith in updates _is_ happening. Both in major updates and in minor.

Again, I am very disappointed I am having to keep an eye on pull requests while this discussion is ongoing.

**Comment by youknowriad** [comment]
> This strategy won’t be applied in Core because it’s very important that we continue to maintain the user trust with auto-updates and there’s no intention in changing that in the future.

I think there's an important misunderstanding here that need to be clarified (as a person who wrote this article). Experimental and unstable APIs were never considered as part of the API, meaning there's no expectation to ensure backward compatibility for this as mentioned in the [documentation](https://developer.wordpress.org/block-editor/contributors/code/coding-guidelines/#experimental-and-unstable-apis) and this is in the wordpress.org website and was discussed in #core-js meeting years ago between core contributors and committers. We've been successfully running with this policy for years now. And to clarify another misunderstanding, this policy applied to code included in Core (not just the Gutenberg plugin) as discussed on these early meetings.  If that's unclear in the current docs, we should make sure to clarify it.

Now, I hope you agree with me that we core committers during these meetings took these decisions with all good faith because we believed that this was the best for the project going forward. 

I understand that you disagree with the policy and that it's your right to raise concerns to potentially improve the policy, especially since no policy is written in stone. I'd suggest that you bring your concerns to one of the future #core-js meetings to be discussed and see where we can go from there.

I think it's very unfair to expect that a concern from a core committer over an existing policy (I guess you didn't think it was one) should block ongoing work.

**Comment by getdave** [comment]
> To continue this discussion, I have to say I am a little annoyed to be told I am making a leap to presume there will be backward compatibility breaks only to discover that https://github.com/WordPress/gutenberg/pull/40052 was intended to be merged last week unless someone objected. https://github.com/WordPress/gutenberg/pull/40052#issuecomment-1130103017
> @Clorith had already objected on the pull request and I was pretty clear in this discussion that I objected to such breaking changed too. It's provocative and very disappointing.

@peterwilsoncc Whilst I appreciate your frustration, I am sorry to see this being perceived as being provocative. At the very least I can assure you it was not deliberately so. 

As the author of that deprecation PR, I have tried to act in good faith to provide maximum communication around the impact of the change. It has been raised in x2 Core Editor chats and numerous folks have done direct outreach around it.

That said, I appreciate you felt you (and perhaps others) [had raised a red flag here](https://github.com/WordPress/gutenberg/pull/40052#issuecomment-1132382728). That wasn't fully clear to me until you commented with a blocking review, others having previously approved. 

I'll hold my hands up and confess I'd fallen behind on the amount of discussion underway in _this_ Issue. Had I been fully up to speed, I would in all likelihood not looked to merge the PR as the (wider) Issue is clearly still under review. I apologise for how that may have come across.

At this point I'm more than happy to take a step back, pause and find a way to resolve this. However, I trust you will appreciate that at no point did I _intend_ to act in a provocative manner. I hope it's clear from my communication and actions that I want the same as you - to ensure we act with the best interests of the community in mind.

**Comment by peterwilsoncc** [comment]
I've created #41265 for the discussion of hard deprecation of formally stable APIs.

> I think it's very unfair to expect that a concern from a core committer over an existing policy (I guess you didn't think it was one) should block ongoing work.

@youknowriad I think this is quite disrespectful to Marius. I shouldn't have to back him up as a second committer to make the point stick. Sure, push back and seek a resolution but put things on hold. I know he and I are not the only Core committers who have this concern. I'm won't claim to be inundated but I am getting messages thanking me for raising these concerns.

As I've mentioned above, the [documentation](https://developer.wordpress.org/block-editor/contributors/code/coding-guidelines/#experimental-and-unstable-apis) is hidden away from theme and plugin developers. By placing it in the contributor handbook, it's only documented to ourselves (by which I mean contributors). 

Unfortunately I can't attend either the editor or JavaScript meetings because they are at midnight at 2AM my time respectively. 

> I hope it's clear from my communication and actions that I want the same as you - to ensure we act with the best interests of the community in mind.

@getdave Yes, I realise this. The reason I am pushing back is to try and restore some of the confidence that has been lost in updating WordPress. 

Breaking a developer API is one thing, but with deprecation procedures running when the editor opens there is also a risk of breaking an owners content.

**Comment by youknowriad** [comment]
> The reason I am pushing back is to try and restore some of the confidence that has been lost in updating WordPress.

I understand but I disagree that the confidence has been lost because of this policy. and If there were a consensus around this, I'd be the first one to call for stopping this practice. Is there anything that shows that the confidence has been lost because of changes to unstable or experimental APIs?

I believe the level of breaking changes that happens in the block editor / Gutenberg are relatively equivalent to the ones that happen in Core php code. The difference is that there's much more changes in the block editor these days because that's where most of the work is happening.

I'd even argue that changes to HTML or CSS (in the admin) (regardless if they come from Gutenberg packages or Core) is causing more issues than changes coming from experimental and unstable APIs. I'm sure you'd agree with me that we can't stop making changes to HTML and CSS because that basically means stop making changes full stop. 

In other words, the experimental and unstable APIs are things that are "exposed" because there's no way around it. We don't want to expose these, it's just that we can't not expose them. It's the same thing for HTML and CSS, any element addition or removal is exposed to the user even if we don't want it.

Our responsibility is to have a clear communication with third-party developpers of what they're safe to use, and things that are unsafe and for things that are unsafe, every time we want to touch them, it's our responsibility to check the impact of these inevitable "breaking changes".

**Comment by Clorith** [comment]
Experimental APIs are only a part of it, and a wider discussion about backwards compatibility and other core concepts need to be had, and I don't think this ticket specifically about experimental APIs is the best place for that, but since it's been brought up, I can summarize a bit here, there's been calls for this discussion in multiple locations over the past year or more, but none has happened yet unfortunately.

Loss of confidence is a mixture between loss of experimental or unstable APIs that have been merged to core (once in core, people will use it, and as noted a few times the warning label is hidden behind the documentation for Gutenberg contributors). and outward facing changes (CSS selectors, markup and similar), and block registrations and dependencies.

One great example is that a core update made blocks suddenly require registration both client and server side, if you want to keep the PHP bit of your block limited, you then need to use `unstable_bootstrapServerSideBlockDefinitions` to avoid errors or missing blocks.

I'm a big advocate for giving users a native-like experience, that means that if core uses an experimental feature, developers also need to use the same experimental features to give a consistent editorial experience.

This is just a few points, there are others, but this does not feel like the correct platform, am very open to a more discoverable place for a wider discussion on this.

**Comment by youknowriad** [comment]
Thanks @Clorith That's great feedback. My disagreement here is about the experimental/unstable APIs discussion. I believe these are independent from the loss of confidence here. 

> I'm a big advocate for giving users a native-like experience, that means that if core uses an experimental feature, developers also need to use the same experimental features to give a consistent editorial experience.

I understand this but I'm sure you agree with me that you need "private" code in any code base. That's what these are about and it's just that we can't make them private. We want to but we can't. It is unrealistic to expect all the code base to be public and available and at the same time commit to backward compatibility. 

Now for the other part of the feedback, I think we should work to solve these.

> One great example is that a core update made blocks suddenly require registration both client and server side, if you want to keep the PHP bit of your block limited, you then need to use unstable_bootstrapServerSideBlockDefinitions to avoid errors or missing blocks.

I'm not personally aware of the particular changes that triggered these forced changes but I agree that we should have definitely ensure backward compatibility here. My understanding when we introduced server side registration of blocks was that it was optional but recommended but not mandatory, maybe in the process we made it so intentionally which we should fix.

> as noted a few times the warning label is hidden behind the documentation for Gutenberg 

As said above, that's a good point and I think we should fix that by making this more prominent, seems like a quick win.

> outward facing changes (CSS selectors, markup and similar)

I agree (especially early after 5.0 when we were not aware of the impact of some changes yet) that we did mistakes here and made some markup changes that we shouldn't have and we should always be careful. Sometimes these changes are inevitable, I believe we learned from this and that recently we're very clear and explicit about these changes. We try to write make/core posts every time there's a need for any, explain the reasons, show paths forward and what we did to mitigate the changes. (The gallery block refactor is a good example here)

**Comment by gziolo** [comment]
I have been playing with the concept of hiding experimental APIs from usage outside of the WordPress codebase controlled by the build process inside the Gutenberg plugin and WordPress core. I opened https://github.com/WordPress/gutenberg/pull/41278 to present the idea we can explore further if that would help prevent leaking experiments in the future. It essentially means we could ensure that plugins and themes can't use experiments with WordPress core as they wouldn't be able to ensure that they provide the secret key that changes (at least) on every WP version bump.

**Comment by Clorith** [comment]
Would it not make more sense that we not include experimentals in core releases, and limit them to the plugin which is intended for testing purposes? Hiding build processes behind secret keys feels like an anti-pattern.

The focus going forward should then be on making any currently experimental features in core stable, give them soft deprecations with a form of "doing it wrong" for JS (I don't know if the JS code has this) like core does for PHP functions already. These should still proxy the request to the stable function call to avoid features from no longer working where needed, and then moving forward not allowing experimentals to pass into a proper release, as by context they are not ready for release?

The reason for soft deprecations for any features already landed in core is essentially because we can not guarantee they're not in use in the wild, nature of the beast, as it were. So anything we've already shipped should be treated as if it were in use.

**Comment by adamziel** [comment]
> Would it not make more sense that we not include experimentals in core releases, and limit them to the plugin which is intended for testing purposes?

@Clorith Some `__experimental` APIs are exported from Gutenberg Package A to be consumed in Gutenberg Package B.  How would you technically remove these API from core releases? IMHO it's not possible and the secret keys come as close as it gets. I'm happy to be wrong here.

**Comment by Clorith** [comment]
So what you're saying is that it's not possible to make the block editor without the use of experimental features? Is it then unreasonable to presume that integrators also need to use experimental features to properly work with the editor, and we should do everything we can not to alienate them?

Yes, this is me playing devils advocate :)

**Comment by adamziel** [comment]
> So what you're saying is that it's not possible to make the block editor without the use of experimental features? 

@Clorith I feel a bit tense and irritated after reading your response because I didn't say that. I asked how would you technically remove these API from core releases after you proposed to do that. I also said that, to me, secrets come close. We do seem to have a few different topics going on in parallel here. I'm happy to talk about the larger need to include the experimental features, but would you mind answering my original question?

**Comment by Clorith** [comment]
My intent was not to irritate, but to flip it around; You're saying that these experimental features are needed to making other components in the block editor work, so then we need to look at it from the position of someone trying to extend the editor, and understand that they are also bound by that same need.

So if we did set a hard block on experimental functions for core moving forward, we would then add a bit more work, because related features would also need to be stabilized before a feature lands. I don't see this as a bad thing necessarily, as it would help ensure that what lands in core is something everyone can rely on as an implementation that can be used, extended, or whatever else one might want to do with it.

Do I have a perfect solution, probably not, I'm just trying to spark a discussion and make sure that we move in the right direction :)

**Comment by dmsnell** [comment]
Adding a note to call for a bit of pragmatism here: we're discussing a lot of ideological values and I get the sense that those risk leading us astray.

In summary of some of the comments above, we don't have to be discussing _if_ developers have lost trust in the API stability because we're specifically talking about a move that _does_ break that trust. In other words, we're talking about removing code that we know many plugins depend on because we can see that in their source code.

Good or bad we expose these APIs and people depend on them and to that matter it's _we_ as the project that haven't adhered to the policy, and not through active intention. It's just how software works that any element that gets exposed becomes part of the public API.

I'm in favor of stabilizing experimental APIs. I'm not in favor of _needlessly_ breaking code that people have come to depend upon. It's not considerate; it's not gaining us anything other than a few bytes that are insignificant in comparison to other changes in bundle sizes.

How long is too long to leave deprecations in place? I don't know, but I'd rather risk leaving them in a little too long while we can still see that a large part of the community is using them. "You shouldn't have done it that way" is never going to be consoling to someone who suddenly starts fielding support calls from people whose content is broken after a WP update.

So for the sake of this issue, I frame it on this question: would you rather suddenly break a lot of code that people have used that we didn't want them to but which we gave them in order to maintain a philosophical purity about a policy we kind of made or would we rather prefer slowing down transitions to communicate to people that we have already slated for certain hooks and functions to be removed and they should use X instead?

Again, there's almost no cost associated to stabilizing without breaking existing code and we win trust and favor. Why are we arguing about intentionally breaking integrations we know are out there and in use? Why would we rather have the editor crash when someone's just trying to write their content than to have an experimental API stick around as a deprecation for _a little while longer_?

**Comment by adamziel** [comment]
> My intent was not to irritate, but to flip it around; 

👍 thanks for clarifying! You're making some good points, it's just a very hot topic and I found that surfacing and labeling the emotions often makes it easier to talk.

> You're saying that these experimental features are needed to making other components in the block editor work, so then we need to look at it from the position of someone trying to extend the editor, and understand that they are also bound by that same need.

@Clorith Are you talking about the stale, long-present __experimental APIs? I do agree. Are you talking about the fresh ones that are actually internal functions that happened to be exposed in public API because of technical limitations of JS? In this case, I disagree. Bear with me, I'll address that in details later in this comment.

That's the culprit I think – we're having many different discussions at the same time here:

* What was actually agreed upon with regard to `__experimental` in the past?
* Does that decision still make sense?
* What to do with the existing `__experimental` APIs?
* How to deal with the future `__experimental` APIs?
* How can we all build an extensible editor?
* Is it okay to internalize the `__experimental` APIs without giving contributors access to them?
* ...probably a few more

No wonder it's confusing and conflicting – that's such a huge discussion scope! I didn't really expect it will go in this direction. I'm not sure whether linear discussion in GitHub issues is the best format to untangle all of that – it's hard to address the abundance of issues that got conflated here. I wonder whether we should use the `discussions` GitHub feature or branch out more issues to spark discussions laser-focused on the specific aspects of that?

To me, the two most actionable problems are:

## Dealing with the **existing** `__experimental` APIs

We're discussing `__experimental` APIs with the following reach:

* APIs used by millions and discoverable via code search
* APIs used by millions and not discoverable via code search
* APIs used by few/none (hard to distinguish from ↑)

These APIs fall into one of two buckers:

* APIs matured and can be stabilized
* APIs that didn't pan out and that we'd like to ditch

The three options we seem to be debating with regard to  `__experimental` APIs that are currently exposed:

* Keep them indefinitely, stabilize some today
* Deprecate them, stabilize some today and remove others after a few releases.
* Stabilize some today, remove others today

There's 18 different paths across these lists, e.g. popular experimental APIs that already matured and we'd like to keep around indefinitely. That's a lot of ground to cover in a single issue! All the downsides and upsides of each option have been already mentioned and it doesn't seem like reiterating them moves the discussion forward anymore.

Since top-down planning seems hard, I echo @mtias and propose we try a discovery/bottom-up route:

1. Create some PRs to stabilize specific APIs
2. Discuss it on a case-by-case basis there. Specific questions to consider:
* What's the cost of hard-removing that specific API?
* What's the cost of soft-deprecating that specific API?
* What's the cost of leaving that specific API indefinitely?

We may quickly find that we're clashing at the strategy level but agree at the tactical level. If so, it would help us find a new alignment on the strategy level.

To @dmsnell's point:

> There's almost no cost associated to stabilizing without breaking existing code and we win trust and favor. Why are we arguing about intentionally breaking integrations we know are out there and in use? Why would we rather have the editor crash when someone's just trying to write their content than to have an experimental API stick around as a deprecation for a little while longer?

This applies to APIs that can be stabilized and I'm happy with such resolution. There may be some APIs intended for removal – I don't think we can blanket-cover them in this discussion so I'd like to follow up on a case-by-case basis.

## Dealing with the **future** `__experimental` APIs

### Experimenting without breaking the internet

The arguments in this debate seem to be:

* Adding and removing public `__experimental` APIs as needed is necessary to develop Gutenberg effectively
* Removing public `__experimental` APIs burdens the extenders and is detrimental to the community

Both arguments hinge on the word "public". 

Notice how there aren't many `__experimental` features on the PHP side, if any. Project philosophy aside, in PHP it's easy to restrict access via `private function x()`. This is how the `WebFonts` API got shipped in WordPress 6.0: it's all encapsulated in closures and private functions so that extenders cannot consume it:

https://github.com/WordPress/wordpress-develop/blob/02414638ce37bb4a58fad2491151fb2fcf4a4e73/src/wp-includes/script-loader.php#L3040-L3052

While the support for WebFonts is there, the API remains an implementation detail to clarify over time.

We need this, but for JavaScript. To me, `__experimental` prefix was an attempt to implement cross-module internal functions using a naming scheme. This discussion tells me it's such an unfortunate fact they've been exposed via `window.wp` and we need to develop a much stricter limitation. @noisysocks proposed one above and @gziolo started implementing it in https://github.com/WordPress/gutenberg/pull/41278. I am happy with it, but if anyone has any other ideas – please, please, please speak out!

### Providing the community with well-defined extension points they need

Restrictions aside, @Clorith also said the following:

> we need to look at it from the position of someone trying to extend the editor, and understand that they are also bound by that same need.

We all want to build an extensible editor here, I don't think anyone is debating this.

I also do agree there's a ton of missing extension points. Just yesterday I've seen a request for the ability to disable specific editor features or block controls or inspector controls. There's a whole lot of good work to do in this area. 

At the same time, I don't view all `__experimental` APIs as extension points:

* `__experimentalLinkControl` certainly is one, let's stabilize it.
* `__experimentalUseQuerySelect` is an internal experiment that we luckily didn't have to export – I'd rather keep private for now and retain the remove it on a whim as needed
* `__experimentalElementButtonClassName` makes button elements possible to manage using Global Styles. The feature is needed, but the implementation may fluctuate wildly. It's not a well-defined extension point yet, but, in all likelihood, it will become one in the future. I'd like to avoid prematurely exposing the API to the extenders even if the feature makes it to WordPress 6.1. It's the same situation as WebFonts (cc @scruffian @hellofromtonya).

So:

* Not all `__experimental` are extension points.
* Feature releases can precede some or all extension points releases.
* The future `__experimental` APIs that are **not** meant to be extension points can be safely internalized

Also, the community would greatly benefit from intentionally exposing the `__experimental` APIs once they mature enough to be useful extension points and remain stable over time.

This is at least three different discussions:

* What's the desired lifecycle of features, extension points, and public APIs in Gutenberg given the technical constraints of JavaScript modules?
* What highly requested extension points are missing today?
* How can we build well-defined extension points as we go to provide the developer community with the exact technical features they need AND with the same ease of working with the editor as core developers have today?

I'd say that's beyond scope of this issue which was meant to figure out a way to stabilize the 280 existing experimental APIs.  I'm more than happy to take it elsewhere, though. Do we have any existing issues about these?

**Comment by dream-encode** [comment]
Hi :wave: I am `davidbaumwald` on [W.org](http://w.org/), and I am one of the Core Committers that privately reached out to @peterwilsoncc to echo my concerns about this. 

I also agree with @Clorith that the plugin is the place to iterate on the editor, not Core.  Regarding this...
>There must be a way to ship them and retain the ability to remove them. If anything that's made public cannot be easily removed, there's no point in maintaining Gutenberg as a separate repository. I now err on the side of keeping the ability to hard-deprecate experimental APIs even if they made it as public to WordPress core.

I think this is the opposite as to how I would assess the problem.  The plugin exists to iterate and stabilize _before_ merging into Core, no different than any other feature plugin.

I disagree with this sentiment...
>I'd even argue that changes to HTML or CSS (in the admin) (regardless if they come from Gutenberg packages or Core) is causing more issues than changes coming from experimental and unstable APIs. I'm sure you'd agree with me that we can't stop making changes to HTML and CSS because that basically means stop making changes full stop.
I don't think anyone is advocating for halting development.  The discussion here is that code that makes it into Core is intended to be stable.  As committers, we merge fixes to existing CSS/HTML but must consider back-compat.

Finally, I think it's clear that there's a huge disconnect between what is considered policy.  My understanding is that any documentation appearing in the [Block Editor Handbook](https://developer.wordpress.org/block-editor) applies only to developing with the editor.  Plugins and Themes also have their own handbook for development, but their APIs that appear in Core must abide by the Core Handbook.

**Comment by akirk** [comment]
I have to admit I haven't read all the comments up until now but I wanted to share an idea that I had around this: Could we by default hide experimental APIs behind a getter like in #41278 and add the Gutenberg version when it was introduced as a variable?

Then we could give experimental APIs a certain number of Gutenberg versions to mature (say 5) and after that it would start to `console.log()` TODO messages. Then the decision could be made to promote them to a non-experimental API and through the getter not break plugins that use it, or give it another version to mature.

I know this might ignore some of the previous discussion, I just wanted to share this idea as a possible path.

**Comment by adamziel** [comment]
After some more discussions, I believe this solution comes close to everyone’s goals:

# Remove the `__experimental` prefix before merging new APIs into WordPress core

WordPress core needs stability and backwards compatibility. Gutenberg needs agility and freedom to remove any `__experimental` API. We won't find a reasonable middle-ground without compromising either of these goals, but we can change the integration process.

**Let's only merge those Gutenberg features into Core that would get accepted as a standalone wordpress-develop PR. No special treatment.**

This way:

* Core can deliver the expected level of Backwards Compatibility
* Gutenberg can retain the freedom to remove the __experimental APIs as needed
* The `__experimental` APIs would get actual audits, not [hand wavy ones](https://github.com/WordPress/gutenberg/issues/39977#event-6904161846)
* [It could save the release leads a lot of work](https://github.com/WordPress/gutenberg/issues/40548)

You all came up with this idea, not me. I'm just facilitating the discussion as we somehow never reached a consensus.

**What if a stable feature depends on an experimental feature?**
Then it isn’t actually stable. Let’s stabilize the dependencies first.

**What about the existing __experimental APIs?**
Most of those already merged to core would get a stable alias. It would preserve BC and shouldn't noticeably affect the bundle size. Some will need a different treatment, let's discuss that case-by-case.

**What if we really need to remove an existing __experimental API that's already in core?**
Do we? If so, let's consider that on a case-by-case basis. There are established Core practices like contacting plugin authors, writing make core posts, preferring soft deprecations and so on. I don’t expect to see many instances of this.

**What if we really need to remove a future stable Gutenberg API after it's released in core?**
Yes, it will happen. Let's acknowledge and embrace it. Some good reasons were already mentioned in this discussion and the future will surprise use with many new ones. Again, let's follow the established core practices. For example, [deprecated.php](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/deprecated.php) shows that removing a function body is sometimes okay as long as the name keeps working.

**How to make this happen? There's no process for that today.**
Block-library already [excludes the experimental blocks from the core build](https://github.com/WordPress/gutenberg/blob/afaa4a62415de1e161477f1fb651102c825dc05f/packages/block-library/src/index.js#L290-L292) and other packages could do the same. We'd also need a simple safeguard such as a pre-commit hook or a CI check to confirm there are no new references to __experimental APIs.

There's also a relevant [discussion about opening wordpress-develop PRs in parallel with Gutenberg PRs, at least for PHP changes](https://github.com/WordPress/gutenberg/issues/40548). Right now Gutenberg PHP code gets blanket-backported to core right before a release, which is far from perfect.

**What are the downsides?**

I see two:

* Some Gutenberg features will get merged into the Core later that they currently would be. The total amount of the development work won't change but the merging timeline will.
* Refactoring Gutenberg APIs will be difficult once they get shipped with core. In reality, that’s already the case.

Is there any other downside that I missed? 

# Alternatives

## Keep not acting on this issue

In this scenario everyone loses:

* Deprecating Gutenberg APIs will lead to tensions and debates.
* Some deprecations will fall through the cracks and get merged, meaning core won't deliver on its Backwards Compatibility promise.
* Both sides of this debate will feel unheard and adversarial.

## Align WordPress and Gutenberg Backwards Compatibility policies

It would make both projects worse-off. The needs are just too different. Personally, I think of Gutenberg as of a WordPress Early Access release channel. Aligning Early Access and Stable wouldn't help either.

## Find a way to ship the `__experimental` APIs with core as "internal" and unavailable to plugin authors

@gziolo explored that in #41278 and it didn't get much traction. There were also some technical challenges. I can't think of any alternative approaches, so it seems like a dead-end.

# Do you agree?

If yes, please react with 👍 to this comment. If not, please elaborate.

**Comment by talldan** [comment]
> Let's only merge those Gutenberg features into Core that would get accepted as a standalone wordpress-develop PR

@adamziel Something wasn't clear to me, so it'd be good to clarify. Is the idea to selectively ship Gutenberg features to core? Or are you proposing that every API is stabilised when a WordPress release approaches?

**Comment by adamziel** [comment]
@talldan The idea is to ensure no new `__experimental` API makes it to core. We'd stabilize those that matured and punt those that didn't. You could call it _selective shipping,_ but that phrase makes me think of messy, manual merges that I'd like to avoid. How about we follow the pattern established in block-library where all experimental blocks are hidden behind a feature flag and thus not included in the core build?

https://github.com/WordPress/gutenberg/blob/afaa4a62415de1e161477f1fb651102c825dc05f/packages/block-library/src/index.js#L290-L292

Perhaps this could be automated further with a webpack plugin to automatically remove the `__experimental` exports from core builds.

**Comment by talldan** [comment]
> How about we follow the pattern established in block-library where all experimental blocks are hidden behind a feature flag and thus not included in the core build?

I think something like that would be needed, but perhaps a bit more evolved. A proper configurable feature flagging system.

To be honest, I think it may be a good idea to consider feature flagging a lot of new work anyway. Whether we like it or not, the Gutenberg plugin has widespread usage in production, and the number of bugs and regressions that are shipped currently is very uncomfortable (at least to me). I don't think the risk of using the plugin is adequately conveyed.

Feature flagging doesn't necessarily mean that all experimental features are disabled by default, but perhaps it could work like this:
- Experimental features are active by default in development builds of Gutenberg, but can be toggled off by the user.
- Experimental features are inactive by default in production builds of Gutenberg, but can be toggled on by the user.
- Features can be promoted or relegated between development or production without re-writing code.
- Experimental features are excluded completely from builds for WordPress core until the code is re-written to remove the feature flag - signifying stability.

**Comment by peterwilsoncc** [comment]
> To be honest, I think it may be a good idea to consider feature flagging a lot of new work anyway. Whether we like it or not, the Gutenberg plugin has widespread usage in production, and the number of bugs and regressions that are shipped currently is very uncomfortable (at least to me). I don't think the risk of using the plugin is adequately conveyed.

True the plugin is used in production. I think this is becoming less common as features get shipped in Core based on anecdotes I am hearing from agencies.

I agree with @adamziel's comment that Gutenberg is _a WordPress Early Access release channel_ and needs a more relaxed attitude to backward compatibility that WordPress Core. Possibly the Early Access section in the readme needs to be earlier and be less ambiguous about not being production ready.

As indicated by my 👍 vote, I support Adam's proposal and think it an excellent way to move forward.

**Comment by Clorith** [comment]
I'd like to echo the sentiment from @peterwilsoncc (and also, huge thanks to @adamziel for summarizing this discussion really well, to what I believe is the right path forward); Let's make it more obvious that the Gutenberg plugin is a alpha/beta-channel for upcoming changes to certain areas of core, instead of hiding things behind feature flags, which some may not even know are there. I use the plugin for experimenting, but I honestly never remember to visit the experiments page to turn on things, so if they're off at any stage, they're not as discoverable as we would like (or that's my perception at least).

And as an agency, I can confirm, we've moved completely away from using the plugin in production. it was a necessity for a while, to get your hands on features that would not be ready for a few core releases, but we've found that we're past that hurdle, the editor as it stands has most of the flexibility an agency needs at least by now.

**Comment by tradesouthwest** [comment]
Gutenberg should not be _anything_ core, let alone "`__experimental`." Gutenberg should not ship with core and should stand on its own; like WooCommerce or any Page builder plugin. Leave "__experimental" to developers, not to the mercy of the community at-large who are having enough trouble trying to understand why the APIs are using a third and a fourth programming language to parse any data of a block; instead of using the PHP and flat javascript of what WP developers already know and trust.

Basically, anything new, conceptually (experimental) or progressive enough as a page builder or ecommerce attachment should stay away from core 101%. 

So my vote is to not promote use of anything interim or temporary into core. Core is for stable. Plugins are for experimental.

**Comment by peterwilsoncc** [comment]
With the introduction of the `@wordpress/experiments` package to make it very, very difficult for theme and plugin developers to use new experimental developer APIs, I think there now exists a robust solution for allowing function signatures that will change.

Thanks @adamziel for figuring out a solution that allows for an early-access channel to co-exist with the stable product.

**Comment by adamziel** [comment]
Yay indeed @peterwilsoncc! Now that there is a way forward, the next steps would be:

#### Tame new experimental APIs

1. Update the handbook with specific guidelines for adding experimental APIs
2. Enable different use-cases via `@wordpress/experimental`
3. Make the unpublished experimental APIs private
4. Add build warnings when new experimental APIs are exported

Related to number 2 above, here's a PR to enable [experimental selectors](https://github.com/WordPress/gutenberg/pull/44521). Other than that there are:

* Private actions – quite similar to the selectors
* Private function arguments – it could take two separate signatures 
* Private React component properties
* Private functions and constants – already supported by `@wordpress/experiments`
* ...anything else?

#### Stabilize the existing experimental APIs

Work through the existing __experimental APIs to add deprecations and provide stable versions when applicable.

**Comment by adamziel** [comment]
Update: Private APIs are now separate from experimental APIs 🎉 

## Private APIs

Private APIs are now shared internally inside the Gutenberg codebase using [the new lock and unlock utilities](https://github.com/WordPress/gutenberg/pull/46131). They can't be imported from the outside. The contributors started using `lock` and `unlock` rapidly and @ntsekouras [coordinated the effort](https://github.com/WordPress/gutenberg/issues/47196) to convert any new `__experimental` exports into private ones for  WordPress 6.2.

I reviewed the `wp/6.2` Gutenberg branch and found only two new __experimental exports: `__experimentalUseSlotFills` and `__experimentalPopoverPositionToPlacement`. This is the lowest number I've seen by a wide margin. There were WordPress releases with 40 new experimental exports, 20 new exports, and even 14, but not less. Two is phenomenal 🎉 

## Experimental APIs

Experimental APIs are like early access feature previews that require community feedback to mature. New experimental APIs should not be merged into WordPress core, but they will continue to be released in the Gutenberg plugin.

## Next steps

This issue had two goals:

* Tame new experimental APIs
* Stabilize the existing experimental APIs

The first goal is now largely achieved. The remaining follow-up work is tracked separately in https://github.com/WordPress/gutenberg/issues/47786. 

As for the second goal, let's track it separately in https://github.com/WordPress/gutenberg/issues/48743. It will require a case-by-case assessment of a few hundred `__experimental` APIs that are in WordPress core today. WordPress 6.2 dealt with a few such exports, and the rest of them will likely be gradually addressed in each future release. Some of them will get stabilized, others will get deprecated, and others may stay indefinitely. The upcoming community summit may yield some answers, as @peterwilsoncc proposed the problem of deprecating JavaScript APIs as a discussion topic.

I really appreciate the discussion we've had here and what it achieved – thank you, everyone!

**Comment by jeffpaul** [comment]
Great work here folks, thanks for all your efforts!
